File: helpers/PythonVersion.py
from ast import parse
import re
import subprocess
from typing import TypedDict

from packaging.version import Version, VERSION_PATTERN

class PipVersion(TypedDict):
	version: str
	filename: str
	downloadUrl: str

class PythonVersion:
	def __init__(self, versionNumber: str, releaseDate: str, downloadUrl: str, installerFilename: str) -> None:
		self.versionNumber = versionNumber
		self.versionInfo = Version(versionNumber)
		self.isStable = not self.versionInfo.is_prerelease
		self.releaseDate = releaseDate
		self.downloadUrl = downloadUrl
		self.installerFilename = installerFilename
		self.pipVersion = self.getPipVersion()

	def __str__(self) -> str:
		return f'Python {self.versionInfo.major}, {self.versionNumber}, released in date {self.releaseDate}'
	
	def __repr__(self) -> str:
		return f'{self.versionInfo.major}, {self.versionNumber}, {self.isStable}, {self.releaseDate}, {self.downloadUrl}'

	def getPipVersion(self) -> PipVersion:
		# fetch the right version from here:
		# https://bootstrap.pypa.io/pip/

		if self.versionInfo >= Version('3.7.0'):
			return {
				'version': 'latest',
				'filename': f'get-pip-latest.py',
				'downloadUrl': f'https://bootstrap.pypa.io/pip/get-pip.py'
			}
		else:
			v = f'{self.versionInfo.major}.{self.versionInfo.minor}'
			return {
				'version': v,
				'filename': f'get-pip-v{v}.py',
				'downloadUrl': f'https://bootstrap.pypa.io/pip/{v}/get-pip.py'
			}


class PythonVersions(TypedDict):
	all: list[str]
	stable: list[str]
	unstable: list[str]
	classes: dict[str, PythonVersion]
	creationDate: int


def getVersionInUse() -> str:
	try:
		output = str(subprocess.check_output(['python', '-V']).strip())
		res = output.split(' ')[1][:-1]
		# print(f'"{res}"')
		return res
	except subprocess.CalledProcessError:
		pass

	return ''

File: internal/PythonVersion/PythonVersion.go
package PythonVersion

import (
	"fmt"

	"github.com/Masterminds/semver/v3"
)

type PipVersion struct {
	Version     string
	Filename    string
	DownloadUrl string
}

type PythonVersion struct {
	VersionNumber     string
	VersionInfo       *semver.Version
	IsStable          bool
	ReleaseDate       string
	ReleaseDateInt    int64
	DownloadUrl       string
	InstallerFilename string
	PipVersion        PipVersion
}

func NewPythonVersion(versionNumber string, releaseDate string, releaseDateInt int64, downloadUrl string, installerFilename string) *PythonVersion {
	v, _ := semver.NewVersion(versionNumber)

	isStableStr := v.Prerelease()
	isStable := false

	if isStableStr == "" {
		isStable = true
	}

	pipVersion := getPipVersion(v)

	return &PythonVersion{
		VersionNumber:     versionNumber,
		VersionInfo:       v,
		IsStable:          isStable,
		ReleaseDate:       releaseDate,
		ReleaseDateInt:    releaseDateInt,
		DownloadUrl:       downloadUrl,
		InstallerFilename: installerFilename,
		PipVersion:        pipVersion,
	}
}

func (pythonVersion *PythonVersion) String() string {
	return fmt.Sprintf("Python %d, version %s, released in date %s", pythonVersion.VersionInfo.Major(), pythonVersion.VersionNumber, pythonVersion.ReleaseDate)
}

func (pythonVersion *PythonVersion) GetPipVersion() PipVersion {
	return pythonVersion.PipVersion
}

func getPipVersion(v *semver.Version) PipVersion {
	s := semver.MustParse("3.7.0")
	if v.LessThan(s) {
		majorMinor := fmt.Sprintf("%d.%d", v.Major(), v.Minor())
		return PipVersion{
			Version:     majorMinor,
			Filename:    fmt.Sprintf("get-pip-v%s.py", majorMinor),
			DownloadUrl: fmt.Sprintf("https://bootstrap.pypa.io/pip/%s/get-pip.py", majorMinor),
		}
	} else {
		return PipVersion{
			Version:     "latest",
			Filename:    "get-pip-latest.py",
			DownloadUrl: "https://bootstrap.pypa.io/pip/get-pip.py",
		}
	}
}
